### PR TITLE
make os_security_group_rule idempotent

### LINKED ISSUE #19610: os_security_group_rule is not idempotent for port_range_min: -1 as advertised in the documentation
_From @stefanandres on July 5, 2016 14:56_

##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

os_security_group_rule
##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.1.0.0
```
##### SUMMARY

When using the example from http://docs.ansible.com/ansible/os_security_group_rule_module.html for '# Another way to create the ping rule' - but use tcp instead of icmp, ansible is not idempotent.
##### STEPS TO REPRODUCE

```
$ cat foo.yml
- hosts: localhost
  become: no
  tasks:
    - name: foo
      os_security_group_rule:
        security_group: default
        protocol: icmp
        port_range_min: -1
        port_range_max: -1
        remote_ip_prefix: 1.1.1.1/32
```

```
$ ansible-playbook foo.yml
PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [foo] *********************************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```

```
$ ansible-playbook foo.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [foo] *********************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Error creating security group rule: Security group rule already exists. Rule id is 5a63c1dc-c3aa-4ef3-9280-b7b7e9e5dbff."}

NO MORE HOSTS LEFT *************************************************************
    to retry, use: --limit @foo.retry

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

This is needed to do more abstraction of security_group_rule, for example:

```
- name: Create security group rules from os_security_rules_ports
  os_security_group_rule:
    security_group: "{{ item.0.security_group | default(item.0.name) }}"
    protocol: "{{ item.0.protocol | default('tcp')  }}"
    port_range_min: "{{ item.0.port_range_min | default('-1') }}"
    port_range_max: "{{ item.0.port_range_max | default('-1') }}"
    remote_ip_prefix: "{{ item.1 }}"
  with_subelements:
    - "{{ os_security_rules_ports }}"
    - remote_ip_prefixes
  when:
    - item.0.port_range_min
    - item.0.port_range_max
    - item.1 is defined
```

With a hash of

```
os_security_rules_ports:
  - name: allow_office
    remote_ip_prefixes:
      - 1.1.1.1/32
      - 2.2.2.2/32
```

will not work, since it will always throw the above error.

_Copied from original issue: ansible/ansible-modules-core#4108_

### PR DESCRIPTION
**Pull request by atosatto**
##### SUMMARY

Allows to define security group rules using the following syntax

```
- os_security_group_rule:
    cloud: mordred
    security_group: foo
    protocol: icmp
    port_range_min: -1
    port_range_max: -1
    remote_ip_prefix: 0.0.0.0/0
```
when protocol is TCP or UDP.

Fixes #19610.

##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
os_security_group_rule

##### ANSIBLE VERSION
```
ansible --version
ansible 2.3.0.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

### CODE REVIEW

**Comment by atosatto** [comment]
Ehi @rcarrillocruz can you please review my PR?

**Comment by ssbarnea** [comment]
I will try to contact core team to get this fixed ASAP, we will also need to backport it to 2.3.x branch after merge.

**Review by emonty — APPROVED** [review]
shipit

**Comment by abadger** [comment]
Merged to devel and cherrypicked to stable-2.4 for the 2.4.0rc3 release and stbale-2.3 for a potential 2.3.3 release.

**Comment by atosatto** [comment]
Thanks! :)

**Pull request merged**
